### PR TITLE
PR: Choose Qt binding to install from `SPYDER_QT_BINDING` environment variable

### LINKED ISSUE #20201: Support for Qt6
I looked through the open and close issues, and could not find one related to support for Qt6.
Currently, Spyder does not support Qt6 bindings: `PyQt6` and `PySide6`: https://github.com/spyder-ide/spyder/blob/e5334d1c1e0b566147739b766d95127d145d0e3d/spyder/requirements.py#L37

Do you plan on adding support/changing the default to a Qt6 binding, and on which timeline?

### PR DESCRIPTION
**Pull request by dpizetta**
## Description of Changes

* [X] Wrote at least one-line docstrings (for any new functions)
* [ ] Added unit test(s) covering the changes (if testable)

* [ ] Included a screenshot or animation (if affecting the UI, see [Licecap](https://www.cockos.com/licecap/))

This is a proposal  to allow the user to choose the Qt binding to be installed when installing with PIP.

for example using PyQt6 (or PySide6 in the future).

[Since the we could not rely on having other imports other than` stdlib `available on `setup.py` ](https://peps.python.org/pep-0517/#build-environment) when building and installing, we could not test if there is a binding already installed. For example, we could use `QtPy` to identify which one was installed, including it to `setup_requires`, and asking for `qtpy.QT_API. But, also without success. 
`
After testing some other ways to do that, I have implemented this version that checks the SPYDER_QT_BINDING environment variable to proper set the requirement during installation. 

We can use like this:

`SPYDER_QT_BINDING='pyqt6' pip install spyder`

This is also compatible with test environments, where you can set the variable. 

Not sure if this useful for `conda` also, I'm not familiar with it. 

I want to hear from you guys, let me know your thoughts.

Links related to setup imports:
- https://github.com/pypa/setuptools/issues/3939
- https://peps.python.org/pep-0517/
- https://github.com/pypa/pip/issues/12030

### Issue(s) Resolved

Part of #20201

### Affirmation

By submitting this Pull Request or typing my (user)name below,
I affirm the [Developer Certificate of Origin](https://developercertificate.org)
with respect to all commits and content included in this PR,
and understand I am releasing the same under Spyder's MIT (Expat) license.

I certify the above statement is true and correct: dpizetta

### CODE REVIEW

**Comment by rear1019** [comment]
> This is a proposal to allow the user to choose the Qt binding to be installed when installing with PIP.

This is something I thought about as well while working on PR #23118. I really like the proposal: Things work by default, users have the option to chose a binding and only a single binding is installed. Other options I came up with have drawbacks:
* Adding a `pyside2/pyqt6/pyside6` section to `extra_require` while leaving `pyqt5` in `install_requires`: Two bindings are installed.
* Remove Qt bindings from `install_requires` and let the packager/end-user handle the installation of a binding (as mentioned here [1]): Not user-friendly.

[1] https://github.com/spyder-ide/spyder/issues/20201#issuecomment-2489869484

File: setup.py [hunk 1]
```diff
@@ -111,6 +111,37 @@ def get_packages():
     packages = get_subpackages(LIBNAME)
     return packages
 
+def get_qt_requirements(qt_requirements, default='pyqt5'):
+    """
+    Return a list of requirements for the Qt binding according to the
+    environment variable SPYDER_QT_BINDING. If this variable is not set
+    or has an unsupported value it defaults to 'pyqt5'.
```
Comment (rear1019) [inline setup.py:118]: Just a nitpick: In my opinion it’s better to throw an error when an invalid/unsupported value is passed.

File: setup.py [hunk 2]
```diff
@@ -111,6 +111,37 @@ def get_packages():
...
+
+    Parameters
+    ----------
+    qt_requirements : dict
+        A dictionary whose keys are supported Qt bindings and whose values are
+        lists of required packages to install for each binding.
+
```
Comment (ccordoba12) [inline setup.py:125]:
```suggestion
```

No need of a blank line here.

File: setup.py [hunk 3]
```diff
@@ -111,6 +111,37 @@ def get_packages():
...
+    default : str
+        Default Qt binding to use if the environment variable is not
+        in qt_requirements. Default is 'pyqt5'.
+
+    Returns
+    -------
+    install_requires : list
+        A list of required packages to install for the given Qt binding.
+    """
+
```
Comment (ccordoba12) [inline setup.py:135]:
```suggestion
```

I think this is unnecessary.

File: setup.py [hunk 4]
```diff
@@ -201,8 +232,16 @@ def run(self):
     cmdclass=CMDCLASS,
 )
 
+# Qt bindings requirements
+qt_requirements = {
+    'pyqt5' : ['pyqt5>=5.15,<5.16', 'pyqtwebengine>=5.15,<5.16'],
+    'pyqt6' : ['pyqt6>=6,<7', 'pyqt6-webengine>=6,<7'],
```
Comment (ccordoba12) [inline setup.py:238]:
```suggestion
    'pyqt6' : ['pyqt6>=6.5,<7', 'pyqt6-webengine>=6.5,<7'],
```

This is our minimal supported version of pyqt6.

**Review by ccordoba12 — COMMENTED** [review]
Thanks for your work on this @dpizetta! Some small suggestions for you, the rest looks good to me.

**Comment by dpizetta** [comment]
Thank you for taking a look! It should be OK after these changes. I may need some help with the AST errors. Thanks

File: setup.py (hunk 1 above)
Comment (dpizetta) [reply setup.py:119]: Raising ValueError when it couldn't find.

**Review by ccordoba12 — APPROVED** [review]
Looks good to me now, thanks @dpizetta!

**Comment by ccordoba12** [comment]
@meeseeksdev please backport to 6.x

**Pull request merged**